### PR TITLE
Fix default args aggregation in StoryStore.

### PR DESCRIPTION
**Pull request by vskh**
Issue: #12540

### TL;DR
I faced a problem causing storyshots to fail with 0 tests found which is caused by undefined/null argument passed into one of the stories.

### Symptoms
Snapshot test run fails with message: 
```
 FAIL  src/storyshots.test.ts
  ● Test suite failed to run

    storyshots found 0 stories

      1 | import testStorySnapshots from "@storybook/addon-storyshots";
      2 | 
    > 3 | testStorySnapshots();
        | ^

      at testStorySnapshots (node_modules/@storybook/addon-storyshots/dist/api/index.js:108:15)
      at Object.<anonymous> (src/stoshots.test.ts:3:1)
```
The error is confusing and does not give any hints as the root cause. Storybook works fine and is not affected.

Sample repo that demonstrates an issue can be found here: https://github.com/vskh/storybook-storyshots-found-0-stories

### Root cause
Passing in explicitly undefined argument to the story will break storyshots stories detection, e.g.:
```
export const Default: Story = args => ...;
Default.args = {
    prop: null
};
```
In StoryStore, ArgTypes inference in this case is returning undefined entry which then fails to decompose the way that default arguments aggregation expects (see the actual fix).

Retrospectively, now that I know the cause it seems to be not a big deal and rather me doing weird thing instead of omitting an argument (omitting makes it work fine) but I spent few hard hours trying to figure out how two packages with seemingly same configuration exhibit different behaviour so I thought to submit a fix if someone ever hits the same.

## What I did

- fixed default arguments aggregation to handle this case
- added unit test

## How to test

`jest lib/client-api/src/story_store.ts`

### CODE REVIEW

File: lib/client-api/src/story_store.ts [hunk 1]
```diff
@@ -451,8 +451,8 @@ export default class StoryStore {
     };
     const defaultArgs: Args = Object.entries(
       argTypes as Record<string, { defaultValue: any }>
-    ).reduce((acc, [arg, { defaultValue }]) => {
-      if (defaultValue) acc[arg] = defaultValue;
+    ).reduce((acc, [arg, argProps]) => {
+      if (argProps && argProps.defaultValue) acc[arg] = argProps.defaultValue;
```
Comment (tmeasday) [inline lib/client-api/src/story_store.ts:455]:
```suggestion
      if (argProps?.defaultValue) acc[arg] = argProps.defaultValue;
```

Also I think `argType` would be a better name for the variable.

**Review by tmeasday — CHANGES_REQUESTED** [review]
So is this issue here that argType inference on `arg: undefined` is itself `argType: undefined`?

Perhaps that itself is the issue here? WDYT @shilman? What should the argType be?

Thanks for tracking it down @vskh. I'm not sure you are wrong to set the arg to `undefined`.

File: lib/client-api/src/story_store.ts (hunk 1 above)
Comment (vskh) [reply lib/client-api/src/story_store.ts]: Thanks, I updated the branch.

Comment (vskh) [reply lib/client-api/src/story_store.ts]: Yes, that is what happening per my understanding. You can have a look at this line: https://github.com/storybookjs/storybook/blob/f286ca2e88ce045216b1cc491ac0ed93d52e0f50/lib/client-api/src/inferArgTypes.ts#L32

**Comment by shilman** [comment]
I wonder if this fix will remove the need for `argType?.defaultValue`: https://github.com/storybookjs/storybook/pull/13029

**Comment by vskh** [comment]
Yes, that is what happening per my understanding. You can have a look at this line:
https://github.com/storybookjs/storybook/blob/f286ca2e88ce045216b1cc491ac0ed93d52e0f50/lib/client-api/src/inferArgTypes.ts#L32

**Comment by vskh** [comment]
Oops, did not see the last comment. Yeah, it should help. I guess my PR is not needed then?
